src/services/crawl-dids/did_processor.service.ts

```diff
@@ -0,0 +1,236 @@
+import { Action, Service } from "@ourparentcenter/moleculer-decorators-extended";
+import { ServiceBroker } from "moleculer";
+import BullableService from "../../base/bullable.service";
+import { DidEventTypes, SERVICE } from "../../common";
+import { calculateDidDeposit } from "../../common/utils/calculate_deposit";
+import { addYearsToDate, formatTimestamp } from "../../common/utils/date_utils";
+
+
+
+interface DidAttribute {
+    key: string;
+    value: string;
+}
+
+interface DidEvent {
+    type: string;
+    did?: string;
+    attributes?: DidAttribute[];
+    [key: string]: any;
+}
+
+interface DidMessageTypes {
+    event_type: string;
+    did?: string;
+    created?: string;
+    modified?: string;
+    deleted_at?: string | null;
+    height?: number;
+    is_deleted?: boolean;
+    exp?: string;
+    deposit?: string;
+    years?: string;
+    id?: string;
+    changes?: any;
+    [key: string]: any;
+}
+
+@Service({
+    name: SERVICE.V1.DidMessageProcessorService.key,
+    version: 1,
+})
+export default class DidMessageProcessorService extends BullableService {
+    constructor(broker: ServiceBroker) {
+        super(broker);
+    }
+
+
+    private computeChanges(
+        oldRec: Record<string, any>,
+        newRec: Record<string, any>
+    ): Record<string, { old: any; new: any }> {
+        const diff: Record<string, { old: any; new: any }> = {};
+        Object.keys(newRec).forEach((key) => {
+            if (newRec[key] !== oldRec[key]) {
+                diff[key] = { old: oldRec[key], new: newRec[key] };
+            }
+        });
+        return diff;
+    }
+
+
+
+    private async saveHistory(didEvent: DidMessageTypes, changes?: any) {
+        const { modified, ...cleanedEvent } = didEvent;
+        await this.broker.call(`${SERVICE.V1.DidHistoryService.path}.save`, {
+            ...cleanedEvent,
+            changes: changes ? JSON.stringify(changes) : null,
+        });
+    }
+
+
+    @Action({ name: "handleDidMessages" })
+    async handleDidMessages(ctx: { params: { listDidTx: DidEvent[] } }) {
+        const { listDidTx } = ctx.params;
+
+        for (const event of listDidTx) {
+            let processedEvent: DidMessageTypes | null = null;
```

Comment:
What do you really mean by `processedEvent` and `DidMessageTypes` here? To me, this type you are calling `DidMessageTypes` is actually the model you are using to store DIDs on the DB.

I guess this comes from the idea of crawling all DID-related events, so in the DB each saved object would simply store each event. But now you are storing here the current DID state, and using the DID History to keep track of the changes.

So in this model I don't see the purpose of storing `event_type`. And maybe there isn't a need of having both `is_deleted` and `deleted_at`, since if the latter is not null I can interpret that the record is deleted, right?